contrib/babelfishpg_tsql/src/pl_handler.c

```diff
@@ -3886,6 +3980,106 @@ call_prev_ProcessUtility(PlannedStmt *pstmt,
 								queryEnv, dest, qc);
 }
 
+/*
+ * Get the oid and acl of a TSQL proc by name. Raises an error if the proc doesn't exist, or if there are
+ * multiple procs with the same name (and different parameters). Also sets isSameFunc based on whether
+ * the found proc is the exact same proc as requested (i.e. the parameters match).
+ */
+static void 
+pltsql_proc_get_oid_proname_proacl(AlterFunctionStmt *stmt, ParseState *pstate, Oid *oid, Acl **acl, bool *isSameFunc)
+{
+	int					spi_rc;
+	char				*funcname, *query;
+	bool				isnull;
+	Oid					schemaOid, funcOid;
+
+	/* Look up the proc */
+	schemaOid = QualifiedNameGetCreationNamespace(stmt->func->objname, &funcname);
+
+	if ((spi_rc = SPI_connect()) != SPI_OK_CONNECT)
+		elog(ERROR, "SPI_connect() failed in pltsql_proc_get_oid_proname_proacl with return code %d", spi_rc);
+
+	query = psprintf("SELECT oid, proacl FROM pg_catalog.pg_proc WHERE proname = '%s' AND pronamespace = %d", funcname, schemaOid);
+	SPI_execute(query, true, 0);
+
+	if (SPI_processed > 1)
+		ereport(ERROR, 
+			(errcode(ERRCODE_AMBIGUOUS_FUNCTION),
+				errmsg("Multiple procedures are defined with the same name and different parameters. Please ensure that there is only one procedure with the target name before calling ALTER PROCEDURE")));
+
+	if (SPI_processed == 0)
+		ereport(ERROR,
+			(errcode(ERRCODE_INVALID_FUNCTION_DEFINITION),
+				errmsg("No existing procedure found with the name for ALTER PROCEDURE")));
+
+	/* exactly one existing procedure with the given name found, retrieve its oid and acl */
+	*oid = DatumGetObjectId(SPI_getbinval(SPI_tuptable->vals[0], SPI_tuptable->tupdesc, 1, &isnull));
+	*acl = DatumGetAclPCopy(SPI_getbinval(SPI_tuptable->vals[0], SPI_tuptable->tupdesc, 2, &isnull));
+
+	if ((spi_rc = SPI_finish()) != SPI_OK_FINISH)
+		elog(ERROR, "SPI_finish() failed in pltsql_proc_get_oid_proname_proacl with return code %d", spi_rc);
+
+	/* now we need to check if the function is exactly the same proc (i.e. the params match as well) or else
+	 * we will run into issues if we try to delete it.
+	 */
+	funcOid = LookupFuncWithArgs(stmt->objtype, stmt->func, true);
+
+	*isSameFunc = OidIsValid(funcOid);
+}
+
+/*
+ * Update the oid and acl of a pg_proc entry given its address
+ */
+static void
+pg_proc_update_oid_acl(ObjectAddress address, Oid oid, Acl acl)
```

Comment:
This function didn't use acl ?
Why we have update acl in the function body ?

Reply:
We need to take the acl from the old procedure and copy it into the new procedure so that the permissions stay the same. I'll modify this function body to do so